File: src/BookIt.Android/app/src/main/java/com/bookit/app/services/BookItApiService.kt
package com.bookit.app.services

import com.bookit.app.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.encodeToString
import kotlinx.serialization.serializer
import java.net.HttpURLConnection
import java.net.URL
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * REST client for the BookIt API.
 * Uses HttpURLConnection + kotlinx.serialization — no external HTTP library needed.
 *
 * Thread-safety: all token mutations and HTTP calls are dispatched to [Dispatchers.IO].
 */
object BookItApiService {

    private const val BASE_URL = "https://api.bookit.app"

    @Volatile private var accessToken: String? = null

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    // -------------------------------------------------------------------------
    // Token management
    // -------------------------------------------------------------------------

    fun setToken(token: String) { accessToken = token }
    fun clearToken()            { accessToken = null  }

    // -------------------------------------------------------------------------
    // Auth
    // -------------------------------------------------------------------------

    suspend fun login(email: String, password: String): AuthResponse =
        post("/api/auth/login", mapOf("email" to email, "password" to password))

    suspend fun register(
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        membershipNumber: String?
    ): AuthResponse {
        val body = buildMap<String, Any?> {
            put("email", email)
            put("password", password)
            put("firstName", firstName)
            put("lastName", lastName)
            put("membershipNumber", membershipNumber)
        }
        return post("/api/auth/register", body)
    }

    // -------------------------------------------------------------------------
    // Tenant
    // -------------------------------------------------------------------------

    suspend fun getTenant(slug: String): TenantResponse =
        get("/api/tenants/$slug")

    // -------------------------------------------------------------------------
    // Appointments
    // -------------------------------------------------------------------------

    suspend fun getAppointments(
        tenantSlug: String,
        from: LocalDateTime,
        to: LocalDateTime
    ): List<AppointmentResponse> {
        val fmt = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss")
        val path = "/api/tenants/$tenantSlug/appointments?from=${from.format(fmt)}&to=${to.format(fmt)}"
        return get(path)
    }

    // -------------------------------------------------------------------------
    // Generic HTTP helpers
    // -------------------------------------------------------------------------

    private suspend inline fun <reified T> get(path: String): T = withContext(Dispatchers.IO) {
        val conn = openConnection(path, "GET")
        val body = conn.inputStream.bufferedReader().readText()
        conn.disconnect()
        json.decodeFromString(body)
    }

    private suspend inline fun <reified T> post(path: String, payload: Map<String, Any?>): T =
        withContext(Dispatchers.IO) {
            val conn = openConnection(path, "POST")
            conn.doOutput = true
            conn.setRequestProperty("Content-Type", "application/json")
            // Encode via kotlinx.serialization to avoid manual escaping issues
            val bodyJson = json.encodeToString(
                serializer<Map<String, JsonElement>>(),
                payload.mapValues { (_, v) ->
                    when (v) {
                        null       -> JsonNull
                        is String  -> JsonPrimitive(v)
                        is Number  -> JsonPrimitive(v)
                        is Boolean -> JsonPrimitive(v)
                        else       -> JsonPrimitive(v.toString())
                    }
                }
            )
            java.io.OutputStreamWriter(conn.outputStream).use { it.write(bodyJson) }
            val statusCode = conn.responseCode
            if (statusCode !in 200..299) {
                val errBody = conn.errorStream?.bufferedReader()?.readText() ?: ""
                conn.disconnect()
                throw ApiException(statusCode, errBody)
            }
            val body = conn.inputStream.bufferedReader().readText()
            conn.disconnect()
            json.decodeFromString(body)
        }

    private fun openConnection(path: String, method: String): HttpURLConnection {
        val url = URL("$BASE_URL$path")
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = method
        conn.setRequestProperty("Accept", "application/json")
        accessToken?.let { conn.setRequestProperty("Authorization", "Bearer $it") }
        conn.connectTimeout = 15_000
        conn.readTimeout    = 30_000
        return conn
    }
}

// ---------------------------------------------------------------------------
// Exception
// ---------------------------------------------------------------------------

class ApiException(val statusCode: Int, val body: String) : Exception() {
    override val message: String
        get() = when (statusCode) {
            401  -> "Invalid credentials. Please check your email and password."
            400  -> "Bad request: $body"
            else -> "Server error ($statusCode)."
        }
}

File: src/BookIt.iOS/BookIt/Services/BookItAPIService.swift
import Foundation

// MARK: - BookIt API Service

/// Async/await HTTP client for the BookIt REST API.
/// Mirror of the C# `BookItApiService` in `BookIt.UI.Shared`.
actor BookItAPIService {

    static let shared = BookItAPIService()

    private let baseURL = URL(string: "https://api.bookit.app")!
    private var accessToken: String?

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        // ASP.NET Core emits camelCase JSON by default; Swift properties are camelCase.
        // No key conversion needed — the names match directly.
        // Dates come as ISO-8601 strings from ASP.NET Core
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        d.dateDecodingStrategy = .custom { decoder in
            let str = try decoder.singleValueContainer().decode(String.self)
            if let date = iso.date(from: str) { return date }
            // fallback without fractional seconds
            let iso2 = ISO8601DateFormatter()
            iso2.formatOptions = [.withInternetDateTime]
            if let date = iso2.date(from: str) { return date }
            throw DecodingError.dataCorruptedError(
                in: try decoder.singleValueContainer(),
                debugDescription: "Cannot decode date: \(str)"
            )
        }
        return d
    }()

    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        // Use default key encoding; ASP.NET Core model binding is case-insensitive
        return e
    }()

    // MARK: Token management

    func setToken(_ token: String) {
        accessToken = token
    }

    func clearToken() {
        accessToken = nil
    }

    // MARK: - Auth

    func login(email: String, password: String) async throws -> AuthResponse {
        struct Body: Encodable { let email: String; let password: String }
        return try await post(path: "/api/auth/login", body: Body(email: email, password: password))
    }

    func register(
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        membershipNumber: String?
    ) async throws -> AuthResponse {
        struct Body: Encodable {
            let email: String
            let password: String
            let firstName: String
            let lastName: String
            let membershipNumber: String?
        }
        return try await post(
            path: "/api/auth/register",
            body: Body(
                email: email,
                password: password,
                firstName: firstName,
                lastName: lastName,
                membershipNumber: membershipNumber
            )
        )
    }

    // MARK: - Tenant

    func getTenant(slug: String) async throws -> TenantResponse {
        return try await get(path: "/api/tenants/\(slug)")
    }

    // MARK: - Appointments

    func getAppointments(
        tenantSlug: String,
        from: Date,
        to: Date
    ) async throws -> [AppointmentResponse] {
        let fmt = DateFormatter()
        fmt.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
        let fromStr = fmt.string(from: from)
        let toStr = fmt.string(from: to)
        let path = "/api/tenants/\(tenantSlug)/appointments?from=\(fromStr)&to=\(toStr)"
        return try await get(path: path)
    }

    // MARK: - Generic HTTP helpers

    private func get<T: Decodable>(path: String) async throws -> T {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw APIError.invalidResponse
        }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        setAuth(on: &request)
        let (data, response) = try await URLSession.shared.data(for: request)
        try validateResponse(response, data: data)
        return try decoder.decode(T.self, from: data)
    }

    private func post<Body: Encodable, T: Decodable>(path: String, body: Body) async throws -> T {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw APIError.invalidResponse
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        setAuth(on: &request)
        request.httpBody = try encoder.encode(body)
        let (data, response) = try await URLSession.shared.data(for: request)
        try validateResponse(response, data: data)
        return try decoder.decode(T.self, from: data)
    }

    private func setAuth(on request: inout URLRequest) {
        if let token = accessToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
    }

    private func validateResponse(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw APIError.httpError(statusCode: http.statusCode, body: body)
        }
    }
}

// MARK: - API Error

enum APIError: LocalizedError {
    case httpError(statusCode: Int, body: String)
    case invalidResponse

    var errorDescription: String? {
        switch self {
        case .httpError(let code, let body):
            if code == 401 { return "Invalid credentials. Please check your email and password." }
            if code == 400 { return "Bad request: \(body)" }
            return "Server error (\(code))."
        case .invalidResponse:
            return "Unexpected response from the server."
        }
    }
}
